server/src/internalClusterTest/java/org/elasticsearch/indices/settings/UpdateSettingsIT.java

```diff
@@ -881,4 +881,44 @@ public void testNoopUpdate() {
         assertNotSame(currentState, clusterService.state());
     }
 
+    public void testAllSettingStringInterned() {
+        final String masterNode = internalCluster().startMasterOnlyNode();
+        final String dataNode = internalCluster().startDataOnlyNode();
+
+        final String index1 = "index-1";
+        final String index2 = "index-2";
+        createIndex(index1, index2);
+        final ClusterService clusterServiceMaster = internalCluster().getInstance(ClusterService.class, masterNode);
+        final ClusterService clusterServiceData = internalCluster().getInstance(ClusterService.class, dataNode);
+        final Settings index1SettingsMaster = clusterServiceMaster.state().metadata().index(index1).getSettings();
+        final Settings index1SettingsData = clusterServiceData.state().metadata().index(index1).getSettings();
+        assertNotSame(index1SettingsMaster, index1SettingsData);
+        assertSame(index1SettingsMaster.get(IndexMetadata.SETTING_INDEX_UUID), index1SettingsData.get(IndexMetadata.SETTING_INDEX_UUID));
+
+        // Create a list of not interned strings to make sure interning setting values works
+        final List<String> queryFieldsSetting = List.of(new String("foo"), new String("bar"), new String("bla"));
+        assertAcked(
+            admin().indices()
+                .prepareUpdateSettings(index1, index2)
+                .setSettings(Settings.builder().putList("query.default_field", queryFieldsSetting))
+        );
+        final Settings updatedIndex1SettingsMaster = clusterServiceMaster.state().metadata().index(index1).getSettings();
+        final Settings updatedIndex1SettingsData = clusterServiceData.state().metadata().index(index1).getSettings();
+        assertNotSame(updatedIndex1SettingsMaster, updatedIndex1SettingsData);
+        assertEqualsAndStringsInterned(queryFieldsSetting, updatedIndex1SettingsMaster);
+        assertEqualsAndStringsInterned(queryFieldsSetting, updatedIndex1SettingsData);
+        assertEqualsAndStringsInterned(queryFieldsSetting, clusterServiceMaster.state().metadata().index(index2).getSettings());
+        assertEqualsAndStringsInterned(queryFieldsSetting, clusterServiceData.state().metadata().index(index2).getSettings());
```

Comment:
Can we also assert that a key is interned? Could just check that the key on index1 is the same as on index2.

Reply:
Sure, I actually just made it assert that all of the keyset is interned strings on these settings now. That was easier (hard to get a reference to a setting key directly) and stricter :)